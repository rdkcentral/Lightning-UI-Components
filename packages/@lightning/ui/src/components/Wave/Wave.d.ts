import lng from '@lightningjs/core';
import type { Base, Color, StylePartial } from '@lightning/ui-core';

export type WaveStyle = {
  color: Color;
  duration: number;
  height: number;
  keyframes: [number, number, number, number];
  lockedRatio: number;
  lockedPaddingRatio: number;
  radius: lng.Tools.CornerRadius;
  repeat: number;
  w: number;
};

export default class Wave extends Base {
  get style(): WaveStyle;
  set style(v: StylePartial<WaveStyle>);

  // tags
  get _Left1(): lng.Component;
  get _Left2(): lng.Component;
  get _Middle(): lng.Component;
  get _Right1(): lng.Component;
  get _Right2(): lng.Component;
}
