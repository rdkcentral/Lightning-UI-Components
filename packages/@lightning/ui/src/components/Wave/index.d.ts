import Wave, { WaveStyle } from './Wave';

export { Wave as default, WaveStyle };
