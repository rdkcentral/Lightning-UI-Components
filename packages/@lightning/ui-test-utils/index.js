export * from './src/lightning-test-renderer';
export * from './src/lightning-test-utils';
export * from './src/lng-test-env';
