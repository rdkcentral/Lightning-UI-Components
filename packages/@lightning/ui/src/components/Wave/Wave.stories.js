import lng from '@lightningjs/core';
import Wave from '.';
import mdx from './Wave.mdx';
import { CATEGORIES } from 'lightning-ui-docs';

export default {
  title: `${CATEGORIES[512]}/Wave`,
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Wave: {
          type: Wave
        }
      };
    }
  };

Basic.args = {
  'toggle animation': true
};

Basic.argTypes = {
  'toggle animation': {
    defaultValue: true,
    control: {
      type: 'boolean'
    },
    description:
      'utilize the startAnimation() and stopAnimation() methods to toggle state',
    table: { defaultValue: { summary: false } }
  }
};

Basic.parameters = {
  argActions: {
    'toggle animation': (shouldAnimate, component) => {
      shouldAnimate
        ? component.tag('Wave').startAnimation()
        : component.tag('Wave').stopAnimation();
      component._refocus();
    }
  }
};
