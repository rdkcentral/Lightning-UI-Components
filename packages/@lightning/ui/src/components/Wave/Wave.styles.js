export const base = theme => ({
  duration: 2,
  keyframes: [0, 0.17, 0.58, 1],
  lockedRatio: 10 / 3,
  lockedPaddingRatio: 5 / 3,
  radius: theme.radius.none,
  repeat: -1,
  w: theme.spacer.sm
});

export const palette = theme => ({
  neutral: { color: theme.color.fillNeutral },
  inverse: { color: theme.color.fillInverse },
  brand: { color: theme.color.fillBrand }
});
