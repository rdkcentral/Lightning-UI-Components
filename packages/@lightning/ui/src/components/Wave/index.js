import Wave from './Wave';

export { Wave as default };
