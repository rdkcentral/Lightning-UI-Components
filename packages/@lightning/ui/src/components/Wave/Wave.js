import lng from '@lightningjs/core';
import { Base } from '@lightning/ui-core';
import * as styles from './Wave.styles';

export default class Wave extends Base {
  static get __componentName() {
    return 'Wave';
  }

  static get __themeStyle() {
    return styles;
  }

  static _template() {
    const mountY = 0.5;
    return {
      Left1: { mountY },
      Left2: { mountY },
      Middle: { mountY },
      Right1: { mountY },
      Right2: { mountY }
    };
  }

  static get tags() {
    return ['Left1', 'Left2', 'Middle', 'Right1', 'Right2'];
  }

  _update() {
    const currentlyPlaying =
      this._hasAnimations() && this._isAnimationPlaying();
    this._basicOptions = {
      repeat: this.style.repeat,
      duration: this.style.duration
    };

    this.stopAnimation();

    this._updateRectangleSize();
    this._updateLines();
    this._updateAnimations();

    if (currentlyPlaying) {
      this.startAnimation();
    }
  }

  _updateRectangleSize() {
    this.h = this.w * this.style.lockedRatio;
    this._padding = this.w * this.style.lockedPaddingRatio;
    this._left2H = this._right1H = this.h * 2;
    this._middleH = this.h * 3;
  }

  _updateLines() {
    this._updateLeft1();
    this._updateLeft2();
    this._updateMiddle();
    this._updateRight1();
    this._updateRight2();
  }

  _updateLeft1() {
    this._Left1.patch(this._generateLine(0, this.h));
  }

  _updateLeft2() {
    this._Left2.patch(
      this._generateLine(this._Left1.x + this._offset, this._left2H)
    );
  }

  _updateMiddle() {
    this._Middle.patch(
      this._generateLine(this._Left2.x + this._offset, this._middleH)
    );
  }

  _updateRight1() {
    this._Right1.patch(
      this._generateLine(this._Middle.x + this._offset, this._right1H)
    );
  }

  _updateRight2() {
    this._Right2.patch(
      this._generateLine(this._Right1.x + this._offset, this.h)
    );
  }

  _updateAnimations() {
    this._updateLeft1Animation();
    this._updateLeft2Animation();
    this._updateMiddleAnimation();
    this._updateRight1Animation();
    this._updateRight2Animation();
  }

  startAnimation() {
    if (this._hasAnimations()) {
      this._left1Animation.play();
      this._left2Animation.play();
      this._middleAnimation.play();
      this._right1Animation.play();
      this._right2Animation.play();
    }
  }

  stopAnimation() {
    if (this._hasAnimations() && this._isAnimationPlaying()) {
      this._left1Animation.pause();
      this._left2Animation.pause();
      this._middleAnimation.pause();
      this._right1Animation.pause();
      this._right2Animation.pause();
    }
  }

  _generateLine(xOffset, lineHeight) {
    return {
      x: xOffset,
      y: this.h,
      texture: lng.Tools.getRoundRect(
        this.w,
        lineHeight,
        this.style.radius,
        0,
        this.style.color,
        true,
        this.style.color
      )
    };
  }

  _generateAnimation(heights) {
    return {
      ...this._basicOptions,
      actions: [
        {
          p: 'h',
          v: this.style.keyframes.reduce((acc, curr, index) => {
            acc[curr] = {
              sm: 0,
              v: Array.isArray(heights[index])
                ? heights[index][0]
                : heights[index]
            };
            if (Array.isArray(heights[index]) && heights[index][1]) {
              acc[curr].sme = heights[index][1];
            }
            return acc;
          }, {})
        }
      ]
    };
  }

  _updateLeft1Animation() {
    const sme = 0.3;
    const heights = [
      [this._h, sme],
      [this._h * 2.5, sme],
      [this._h / 2, sme],
      [this._h, sme]
    ];
    this._left1Animation = this._Left1.animation(
      this._generateAnimation(heights)
    );
  }

  _updateLeft2Animation() {
    const heights = [
      this._left2H,
      this._left2H / 8,
      this._left2H * 1.5,
      this._left2H
    ];
    this._left2Animation = this._Left2.animation(
      this._generateAnimation(heights)
    );
  }

  _updateMiddleAnimation() {
    const sme = 0.5;
    const heights = [
      this._middleH,
      [this._middleH / 2, sme],
      [this._middleH / 3, sme],
      this._middleH
    ];
    this._middleAnimation = this._Middle.animation(
      this._generateAnimation(heights)
    );
  }

  _updateRight1Animation() {
    const sme = 0.5;
    const heights = [
      this._right1H,
      [this._right1H / 2, sme],
      [this._right1H * 1.25, sme],
      this._right1H
    ];
    this._right1Animation = this._Right1.animation(
      this._generateAnimation(heights)
    );
  }

  _updateRight2Animation() {
    const sme = 0.3;
    const heights = [
      [this.h, sme],
      [this.h * 3, sme],
      [this.h * 1.5, sme],
      [this.h, sme]
    ];
    this._right2Animation = this._Right2.animation(
      this._generateAnimation(heights)
    );
  }

  _hasAnimations() {
    return (
      this._left1Animation &&
      this._left2Animation &&
      this._middleAnimation &&
      this._right1Animation &&
      this._right2Animation
    );
  }

  _isAnimationPlaying() {
    return (
      this._left1Animation.isPlaying() &&
      this._left2Animation.isPlaying() &&
      this._middleAnimation.isPlaying() &&
      this._right1Animation.isPlaying() &&
      this._right2Animation.isPlaying()
    );
  }

  get _offset() {
    return this._padding + this.w;
  }
}
