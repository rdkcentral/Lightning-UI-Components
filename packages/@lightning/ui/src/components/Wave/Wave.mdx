import { Canvas, Story } from '@storybook/addon-docs';
import Wave from '.';

# Wave

This component takes the wave icon and animates each line individually.

## Source

https://github.comcast.com/Lightning/lightning-ui/packages/@lightning/ui-core/src/components/Wave/Wave.js

## Usage

`Wave` can be used without passing in any additional values and simply toggling the animation via stop and start.

```js
import { Wave } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Wave: {
        type: Wave
      }
    };
  }
}
```

<Canvas>
  <Story id="utilities-wave--wave" />
</Canvas>

### Sizing

`Wave` height dynamically adjust over the course of the animation. You can set the 'initial' Wave height via this.h.

## API

### Style Properties

| name               | type             | description                                                          |
| ------------------ | ---------------- | -------------------------------------------------------------------- |
| duration           | number           | how long the animation should last                                   |
| keyframes          | array of numbers | sets the animation stops for each line                               |
| lockedRatio        | number           | ratio to maintain for line height when adjusting the width           |
| lockedPaddingRatio | number           | ratio to maintain for spacing between lines when adjusting the width |
| radius             | number           | radius to apply to individual lines                                  |
| repeat             | number           | how many times to repeat the line animations (-1 for infinite)       |

### Methods

#### startAnimation(): void

Start the animation of the circles.

#### stopAnimation(): void

Stops the circles animation. When in the stopped state, the component will fallback to all three circles stacked on top of each other.
