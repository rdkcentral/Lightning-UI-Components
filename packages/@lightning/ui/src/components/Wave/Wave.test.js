import { makeCreateComponent } from '@lightning/ui-test-utils';
import Wave from '.';
const createWave = makeCreateComponent(Wave, {
  color: 'rgba(256, 256, 256, 1)',
  h: 20
});

describe('wave', () => {
  // eslint-disable-next-line no-unused-vars
  let wave, testRenderer;

  beforeEach(() => {
    [wave, testRenderer] = createWave();
  });

  afterEach(() => {
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });
});
