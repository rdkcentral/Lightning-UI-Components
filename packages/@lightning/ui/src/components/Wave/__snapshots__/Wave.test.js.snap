// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wave renders 1`] = `
{
  "Component": {
    "Left1": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Left1",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Left1",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 20,
      "zIndex": 0,
    },
    "Left2": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Left2",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Left2",
      ],
      "texture": {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 16,
      "y": 20,
      "zIndex": 0,
    },
  },
}
`;
